mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,368 @@
-                                                                     * connection
-                                                                     * ids from the
-                                                                     * agent, to be read in the
-                                                                     * 'accept'
-                                                                     * call */
-    pending_data: Mutex<HashMap<ConnectionId, TCPBuffer>>, /* Used to store data that arrived
-                                                            * before its
-                                                            * connection was opened. When the
-                                                            * connection is
-                                                            * later opened, pending_data is read
-                                                            * and
-                                                            * emptied. */
-}
-
-impl Default for Sockets {
+impl Default for SocketState {
     fn default() -> Self {
-        Self {
-            new_sockets: Mutex::new(HashMap::new()),
-            connections: Mutex::new(MultiMap::new()),
-            data: Mutex::new(MultiMap::new()),
-            connection_queues: Mutex::new(HashMap::new()),
-            pending_data: Mutex::new(HashMap::new()),
-        }
+        SocketState::Initialized
     }
 }
 
-impl Sockets {
-    pub fn create_socket(&self) -> SockFd {
-        let (write_socket, read_socket) = socketpair_stream().unwrap();
-        let read_fd = read_socket.as_raw_fd();
-        let socket = Socket {
-            read_fd,
-            read_socket,
-            write_socket,
-        };
-
-        self.new_sockets.lock().unwrap().insert(read_fd, socket);
+#[derive(Debug)]
+struct Socket {
+    fd: RawFd,
+    domain: c_int,
+    type_: c_int,
+    protocol: c_int,
+    pub state: SocketState,
+}
 
-        read_fd
+impl PartialEq for Socket {
+    fn eq(&self, other: &Self) -> bool {
+        self.fd == other.fd
     }
+}
 
-    pub fn convert_to_connection_socket(&self, sockfd: SockFd, address: SocketAddr) {
-        let mut sockets = self.new_sockets.lock().unwrap();
-
-        // let mut sockets = self.connections.lock().unwrap();
-        if let Some(socket) = sockets.remove(&sockfd) {
-            // socket.port = port;
-            self.connections.lock().unwrap().insert(
-                sockfd,
-                address.port(),
-                ConnectionSocket {
-                    read_fd: socket.read_fd,
-                    read_socket: socket.read_socket,
-                    write_socket: socket.write_socket,
-                    address,
-                    state: ConnectionState::Bound,
-                },
-            );
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-        }
-    }
+impl Eq for Socket {}
```

Comment:
Why is this necessary?

Reply:
that's how you impl Eq which is required for HashSet.